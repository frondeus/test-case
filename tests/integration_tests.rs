#![cfg(test)]

use insta::with_settings;
use itertools::Itertools;
use std::env;
use std::path::PathBuf;
use std::process::Command;

macro_rules! run_integration_test {
    ($case_name:expr, $cmd:expr) => {
        with_settings!({snapshot_path => get_snapshot_directory()}, {
            let subcommand = Command::new("cargo")
                .current_dir(PathBuf::from("tests").join("integration_test_cases").join($case_name))
                .args(&[$cmd])
                .output()
                .expect("Failed to spawn cargo subcommand");

            let mut output = String::new();
            output.push_str(String::from_utf8_lossy(&subcommand.stdout).as_ref());
            output.push_str(String::from_utf8_lossy(&subcommand.stderr).as_ref());

            let output = sanitize_lines(output);

            insta::assert_display_snapshot!(output);
        })
    };
    ($case_name:expr) => {
        run_integration_test!($case_name, "test")
    }
}

fn get_snapshot_directory() -> String {
    PathBuf::from("snapshots")
        .join(env::var("SNAPSHOT_DIR").unwrap_or_else(|_| "rust-stable".to_string()))
        .to_str()
        .unwrap()
        .to_string()
}

fn sanitize_lines(s: String) -> String {
    let mut s = s
        .lines()
        .filter(|line| {
            !line.contains("note")
                && !line.contains("error")
                && !line.contains("waiting")
                && !line.contains("Finished")
                && !line.contains("Compiling")
                && !line.contains("termination value with a non-zero status code")
                && !line.contains("Running unittests")
                && !line.contains("Running target")
                && !line.is_empty()
        })
        .map(|line| line.replace('\\', "/"))
        .map(|line| line.replace(".exe", ""))
        .collect::<Vec<_>>();

    s.sort_unstable();

    s.into_iter().join("\n")
}

#[test]
fn basic() {
    run_integration_test!("basic")
}

#[test]
fn async_tests() {
    run_integration_test!("async_tests")
}

#[test]
fn test_item_reuse() {
    run_integration_test!("test_item_reuse")
}

#[test]
fn test_item_reuse_run() {
    run_integration_test!("test_item_reuse", "run")
}

#[test]
fn test_result() {
    run_integration_test!("result")
}
