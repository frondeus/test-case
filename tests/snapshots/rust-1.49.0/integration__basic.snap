---
source: tests/integration_tests.rs
assertion_line: 65
expression: output
---
      panic message: `"It has to panic"`,
    test_cases::extended_pattern_matching_result::err_should_fail
    test_cases::extended_pattern_matching_result::ok_should_fail
    test_cases::panicing::_expects_panicking_some_this_should_fail_
 expected substring: `"This should fail"`
---- test_cases::extended_pattern_matching_result::err_should_fail stdout ----
---- test_cases::extended_pattern_matching_result::ok_should_fail stdout ----
---- test_cases::panicing::_expects_panicking_some_this_should_fail_ stdout ----
failures:
failures:
running 98 tests
test result: FAILED. 91 passed; 3 failed; 4 ignored; 0 measured; 0 filtered out
test test_cases::abs_tests::returns_0_for_0 ... ok
test test_cases::abs_tests::returns_given_number_for_positive_input ... ok
test test_cases::abs_tests::returns_opposite_number_for_non_positive_input ... ok
test test_cases::arg_expressions::_2_4_6_to_string_expects ... ok
test test_cases::bar::_expects_string_default_ ... ok
test test_cases::basic_test::_1_expects ... ok
test test_cases::combinators::_0_3_expects_complex_gt_0_0_and_lt_1_0_or_gt_1_2 ... ok
test test_cases::combinators::_0_7_expects_complex_gt_0_0_and_lt_1_0_or_gt_1_2 ... ok
test test_cases::combinators::_1_0_expects_complex_gt_0_0_and_lt_5_0 ... ok
test test_cases::combinators::_1_0_expects_complex_gt_0_0_or_lt_0_0 ... ok
test test_cases::combinators::_1_0_expects_complex_gt_0_0_or_lt_1_5_and_lt_2_0 ... ok
test test_cases::combinators::_2_0_expects_complex_gt_0_0_or_lt_0_0 ... ok
test test_cases::combinators::_2_0_expects_complex_gt_0_0_or_lt_0_0_and_lt_1_0 ... ok
test test_cases::combinators_with_arrays::vec_1_2_3_expects_complex_contains_1_and_contains_2_and_contains_in_order_2_3_ ... ok
test test_cases::combinators_with_arrays::vec_1_2_3_expects_complex_contains_1_and_contains_3_or_contains_5 ... ok
test test_cases::combinators_with_arrays::vec_1_2_3_expects_complex_contains_1_or_contains_4 ... ok
test test_cases::combinators_with_arrays::vec_1_2_3_expects_complex_contains_1_or_contains_4_and_contains_2 ... ok
test test_cases::combinators_with_arrays::vec_1_2_3_expects_complex_contains_6_and_contains_7_or_contains_1 ... ok
test test_cases::combinators_with_arrays::vec_1_2_3_expects_complex_contains_6_and_contains_7_or_contains_1_and_contains_in_order_1_2_3_ ... ok
test test_cases::complex_tests::almost_eq1 ... ok
test test_cases::complex_tests::almost_eq2 ... ok
test test_cases::complex_tests::eq1 ... ok
test test_cases::complex_tests::eq2 ... ok
test test_cases::complex_tests::geq1 ... ok
test test_cases::complex_tests::geq2 ... ok
test test_cases::complex_tests::gt1 ... ok
test test_cases::complex_tests::gt2 ... ok
test test_cases::complex_tests::leq1 ... ok
test test_cases::complex_tests::leq2 ... ok
test test_cases::complex_tests::lt1 ... ok
test test_cases::complex_tests::lt2 ... ok
test test_cases::const_in_arg::this_is_desc_not_an_argument ... ok
test test_cases::contains_tests::vec_1_2_3_4_expects_complex_contains_1 ... ok
test test_cases::contains_tests::vec_1_2_3_4_expects_complex_contains_in_order_3_4_ ... ok
test test_cases::create_path::_cargo_toml_expects_complex_path_path ... ok
test test_cases::create_path::_src_lib_rs_expects_complex_path_file ... ok
test test_cases::create_path::long_dir ... ok
test test_cases::create_path::short_dir ... ok
test test_cases::divide_by_zero_f64_with_lambda::_0_0_expects_with_v_f64_assert_v_is_nan_ ... ok
test test_cases::divide_by_zero_f64_with_lambda::_1_0_expects_with_v_f64_assert_v_is_infinite_ ... ok
test test_cases::extended_pattern_matching_result::err_should_fail ... FAILED
test test_cases::extended_pattern_matching_result::ok_should_fail ... FAILED
test test_cases::extended_pattern_matching_result::simpleenum_var1_expects_matching_ok_e_e_simpleenum_var1 ... ok
test test_cases::extended_pattern_matching_result::simpleenum_var2_expects_matching_err_e_e_var2_ ... ok
test test_cases::fancy_addition::some_2_3_some_4_expects_2_3_4 ... ok
test test_cases::fancy_addition::some_2_some_3_expects_5 ... ok
test test_cases::fancy_addition::treats_none_as_0 ... ok
test test_cases::impl_trait::_foo_expects ... ok
test test_cases::in_parens::_2_0_expects_complex_eq_2_0 ... ok
test test_cases::inconclusives::_expects_inconclusive_ ... ignored
test test_cases::inconclusives::ignore_keyword ... ignored
test test_cases::inconclusives::inconclusive_test ... ignored
test test_cases::inconclusives::test_is_not_ran ... ignored
test test_cases::keyword_test::true_expects ... ok
test test_cases::leading_underscore_in_test_name::_dummy_expects ... ok
test test_cases::lowercase_test_name::_dummy_code_expects ... ok
test test_cases::multiple_test_cases::_2_expects ... ok
test test_cases::multiple_test_cases::_4_expects ... ok
test test_cases::multiplication_tests::when_both_operands_are_negative ... ok
test test_cases::multiplication_tests::when_both_operands_are_possitive ... ok
test test_cases::multiplication_tests::when_operands_are_swapped ... ok
test test_cases::name::_2nd_test ... ok
test test_cases::name::_4_5_expects ... ok
test test_cases::name::test_1_8_9 ... ok
test test_cases::name::test_3_6_9 ... ok
test test_cases::nested::nested_test_case::_1_1_expects ... ok
test test_cases::nested::using_fn_from_super::_20_22_expects ... ok
test test_cases::nested::using_fn_from_super::_42_expects ... ok
test test_cases::not_complex::_1_0_expects_complex_not_almost_2_1_p_0_01 ... ok
test test_cases::not_complex::_1_0_expects_complex_not_eq_2_5 ... ok
test test_cases::not_contains_tests::vec_1_2_3_4_expects_complex_not_contains_5 ... ok
test test_cases::not_contains_tests::vec_1_2_3_4_expects_complex_not_contains_in_order_3_2_ ... ok
test test_cases::not_path::_cargo_toml_parse_unwrap_expects_complex_not_path_dir ... ok
test test_cases::not_path::_cargo_yaml_parse_unwrap_expects_complex_not_path_path ... ok
test test_cases::not_path::_src_parse_unwrap_expects_complex_not_path_file ... ok
test test_cases::panicing::_expects_panicking_some_it_has_to_panic_ ... ok
test test_cases::panicing::_expects_panicking_some_this_should_fail_ ... FAILED
test test_cases::panics_without_value::_expects_panicking_none ... ok
test test_cases::pattern_matching_result::simpleenum_var2_expects_matching_simpleenum_var2_ ... ok
test test_cases::pattern_matching_result_fails::simpleenum_var1_expects_matching_simpleenum_var2_ ... ok
test test_cases::power_of_two_with_using::_1_expects_use_assert_is_power_of_two ... ok
test test_cases::power_of_two_with_using::_2_expects_use_crate_test_cases_assert_is_power_of_two ... ok
test test_cases::power_of_two_with_using::_4_expects_use_some_mod_assert_is_power_of_two ... ok
test test_cases::pretty_assertions_usage::_1_expects_use_wrapped_pretty_assert_1_ ... ok
test test_cases::qualified_attribute::first_test ... ok
test test_cases::qualified_attribute::second_test ... ok
test test_cases::result::_2_expects_4 ... ok
test test_cases::result::_4_expects_8 ... ok
test test_cases::result_and_name::_4_5_expects_9 ... ok
test test_cases::result_and_name::test_no_1 ... ok
test test_cases::result_expresion_with_name::test_result_expression ... ok
test test_cases::result_expression::_2_2_expects_2_2 ... ok
test test_cases::result_which_panics::_2_2_expects_2_3 ... ok
test test_cases::result_with_mod_sep::_42_expects_std_string_string_new_ ... ok
test test_cases::test_generics::source1_expects_1 ... ok
test test_cases::test_generics::source2_expects_2 ... ok
test test_cases::test_impl::source1_expects_1 ... ok
test test_cases::test_impl::source2_expects_2 ... ok
thread 'test_cases::extended_pattern_matching_result::err_should_fail' panicked at 'Expected `Err(e) if e == "var1"` found Err("var2")', src/lib.rs:153:5
thread 'test_cases::extended_pattern_matching_result::ok_should_fail' panicked at 'Expected `Ok(e) if e == SimpleEnum :: Var2` found Ok(Var1)', src/lib.rs:153:5
thread 'test_cases::panicing::_expects_panicking_some_this_should_fail_' panicked at 'It has to panic', src/lib.rs:174:9
